components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java

```diff
@@ -519,15 +521,25 @@ private void validateAddOrganizationRequest(Organization organization) throws Or
     private void validateOrgNameUniquenessAmongSiblings(String parentOrgId, String organizationName)
             throws OrganizationManagementException {
 
-        boolean hasSiblingWithSameName =
-                organizationManagementDAO.getChildOrganizations(parentOrgId.trim(), false).stream().anyMatch(
-                        basicOrganization -> StringUtils.equals(basicOrganization.getName(), organizationName.trim()));
-        if (hasSiblingWithSameName) {
+        if (organizationManagementDAO.isSiblingOrganizationExistWithName(organizationName, parentOrgId)) {
             throw handleClientException(ERROR_CODE_SAME_ORG_NAME_ON_IMMEDIATE_SUB_ORGANIZATIONS_OF_PARENT_ORG,
                     organizationName, parentOrgId);
         }
     }
 
+    private void validateOrgNameUniqueness(String parentOrgId, String organizationName)
+            throws OrganizationManagementException {
+
+        int depthOfChildOrganization = organizationManagementDAO.getOrganizationDepthInHierarchy(parentOrgId) + 1;
+        if (!isSubOrganization(depthOfChildOrganization)) {
+            validateOrgNameUniquenessAmongSiblings(parentOrgId, organizationName);
+            return;
+        }
+        if (organizationManagementDAO.isChildOrganizationExistWithName(organizationName, parentOrgId)) {
```

Comment:
Is this logic correct?
It checks only the same name is taken by any other organization under the tree of creating the org's parent. but the sub tree might be starting from above level

Let's think 
A->B->C->D->E are the hierarchy . sub-org starting level is 1(which means level B)
If I try to add org named D pointing B as the parent -> it will be failed
If I try to add an org named C as a sub-org of E -> It is possible.

But we want to have unique names for each sub-organizations under B

Reply:
Yes. the logic is wrong as mentioned, before the Line 538, the root org Id should be resolved from the parent orgId.
Will resolve the raised concern here after proper testing

Reply:
Introduced a logic to derive the root org from the list of ancestor orgs. The root org is the level before org of the defined sub-organization starting level. 
Unit test case is added to verify the failure of sub org additions of `C as a sub-org of E` as mentioned in the comment

Fixed by - https://github.com/wso2-extensions/identity-organization-management-core/pull/65/commits/a0a69184239b61c5b79cad296fbc282a0a0b380c